Auto register resource providers

mcardosos opened the pull request:
This would also need to add more generated code, it would look  like this...
```go
func (client TopicsClient) CreateOrUpdateSender(req *http.Request) (*http.Response, error) {
	return autorest.SendWithSender(client,
		req,
		azure.RegisterResourceProvider(),
		autorest.DoRetryForStatusCodes(client.RetryAttempts, client.RetryDuration, autorest.StatusCodesForRetry...),
		azure.DoPollForAsynchronous(client.PollingDelay),
	)
}
```
PTAL @jhendrixMSFT @marstr

marstr commented:
Before I dive into the code too much here, I want to understand how it relates to this method:

[azure-sdk-for-go/arm/resources/resources/providers.go#L252 @ 7362b8b](https://github.com/Azure/azure-sdk-for-go/blob/7362b8bf379bde4d1e26e91ee424ea6c5d8796d3/arm/resources/resources/providers.go#L252)

Is it a case of circular references getting in our way that we wouldn't reuse this?

mcardosos commented:
I would prefer not to reference the SDK in the runtime. It would also be a circular reference.

jhendrixMSFT commented:
@marstr brings up a good point, if we're already generating code that does the work then should we adpot it instead.  Obviously we can't reference the Go SDK due to circular dependencies, but perhaps it makes sense to copy/paste this code instead (i.e. would it set us up better for future servicing of this code).

mcardosos commented:
I copied and pasted code, and left out what made not so much sense in this context (like unmarshaling the response body)

mcardosos commented:
After fixing stuff, generated code would look like this...
``` go
func (client TopicsClient) CreateOrUpdateSender(req *http.Request) (*http.Response, error) {
	return autorest.SendWithSender(client,
		req,
		azure.RegisterResourceProvider(),
		azure.DoPollForAsynchronous(client.PollingDelay),
	)
}
```

mcardosos commented:
This rework with the Azure specific client no longer requires changes on senders in generated code. But it will require changes in the clients file.
``` diff
// Azure EventGrid Management Client
package eventgrid

import (
	"github.com/Azure/go-autorest/autorest"
+	"github.com/Azure/go-autorest/autorest/azure"
)

const (
	// DefaultBaseURI is the default URI used for the service Eventgrid
	DefaultBaseURI = "https://management.azure.com"
)

// ManagementClient is the base client for Eventgrid.
type ManagementClient struct {
-	autorest.Client
+	azure.Client
	BaseURI        string
	SubscriptionID string
}

// New creates an instance of the ManagementClient client.
func New(subscriptionID string) ManagementClient {
	return NewWithBaseURI(DefaultBaseURI, subscriptionID)
}

// NewWithBaseURI creates an instance of the ManagementClient client.
func NewWithBaseURI(baseURI string, subscriptionID string) ManagementClient {
	return ManagementClient{
-		Client:         autorest.NewClientWithUserAgent(UserAgent()),
+		Client: azure.Client{
+			Client: autorest.NewClientWithUserAgent(UserAgent()),
+		},
		BaseURI:        baseURI,
		SubscriptionID: subscriptionID,
	}
}
```

PTAL @jhendrixMSFT

jhendrixMSFT commented on autorest/azure/sender.go:
```diff
@@ -0,0 +1,187 @@
+	return "", errors.New("provider was not found in the response")
+}
+
+func register(azc Client, s autorest.Sender, originalReq *http.Request, re RequestError) error {
```
I would turn this into a method where (azc Client) is the receiver.  Also doesn't the client implement autorest.Sender?  If so then it seems you could remove this param.

mcardosos replied:
I need the client to get authorization again

jhendrixMSFT replied:
Sorry I meant remove the autorest.Sender param.

mcardosos replied:
I use that one for the test :(

jhendrixMSFT commented on autorest/azure/client.go:
```diff
@@ -0,0 +1,62 @@
+}
+
+// AzurePreparation is a prepare decorator that adds a User Agent and Authorization headers to the request
+func (azc Client) AzurePreparation() autorest.PrepareDecorator {
```
I don't think this needs to be exported does it?

mcardosos replied:
Fixed

jhendrixMSFT commented on autorest/azure/sender.go:
```diff
@@ -0,0 +1,187 @@
+	"github.com/Azure/go-autorest/autorest"
+)
+
+// DoRetryForStatusCodes is the Azure specific implementation for retring requests
```
I would update this comment to mention that it will automatically attempt to register the resource provider.

mcardosos replied:
Fixed

mcardosos commented:
With this changes, we can use the clients the same as before (the client only has a different type)
``` go
eventgridClient.RetryDuration = time.Second
eventgridClient.RetryAttempts = 3
eventgridClient.PollingDelay = time.Second * 10
eventgridClient.PollingDuration = time.Minute * 5
resp, errChan := eventgridClient.CreateOrUpdate(groupName, name, eventgrid.Topic{
	Location: to.StringPtr("westcentralus"),
}, nil)
```
This is valid code with and without the change mentioned above to generated code.

jhendrixMSFT commented on autorest/azure/sender.go:
```diff
@@ -0,0 +1,189 @@
+					}
+
+					if re.ServiceError != nil && re.ServiceError.Code == "MissingSubscriptionRegistration" {
+						azc.Sender = s
```
Presumably for non-test scenarios azc.Sender and s are identical?  It would be good to comment why you need to do this assignment.

mcardosos replied:
They are not identical, but they contain the same information

mcardosos commented:
Reworked code to get rid of Azure client. Very important change to note is I removed the retries from the `Do` implementation for `autorest.Client`. I have also fixed the retriable requests bug.
Generated code would look like this, there are no more breaking changes:
``` diff
func (client TopicsClient) CreateOrUpdateSender(req *http.Request) (*http.Response, error) {
	return autorest.SendWithSender(client,
		req,
+		azure.RegisterResourceProvider(client.Client),
		azure.DoPollForAsynchronous(client.PollingDelay),
	)
}
```
PTAL @jhendrixMSFT @marstr

jhendrixMSFT commented on autorest/azure/rp.go:
```diff
@@ -0,0 +1,195 @@
+)
+
+// RegisterResourceProvider  tries to register the resource provider in case it is unregistered
+func RegisterResourceProvider(client autorest.Client) autorest.SendDecorator {
```
Since this is replacing DoRetryForStatusCodes() and augmenting it with registration I would rename this to something that reflects that, maybe DoRetryWithRegistration() or something along those lines.

mcardosos replied:
Fixed

jhendrixMSFT commented on autorest/azure/rp.go:
```diff
@@ -0,0 +1,195 @@
+						return resp, fmt.Errorf("failed auto registering Resource Provider: %s", err)
+					}
+				}
+				err = rr.Prepare()
```
I think this code and the lines below it can be removed, refactoring the code starting at line 19 into a loop.

mcardosos replied:
Fixed

jhendrixMSFT commented on autorest/azure/rp.go:
```diff
@@ -0,0 +1,190 @@
+func DoRetryWithRegistration(client autorest.Client) autorest.SendDecorator {
+	return func(s autorest.Sender) autorest.Sender {
+		return autorest.SenderFunc(func(r *http.Request) (resp *http.Response, err error) {
+			shouldRetry := true
```
I think the var shouldRetry can be removed.

jhendrixMSFT commented on autorest/azure/rp.go:
```diff
@@ -0,0 +1,190 @@
+				}
+				shouldRetry = false
+
+				if resp.StatusCode == http.StatusConflict {
```
After removing the shouldRetry you can change this to exit early on the non-conflicting status.
```go
if resp.StatusCode != http.StatusConflict {
    return resp, err
}
var re RequestError
...
```

mcardosos replied:
done :3

marstr commented on autorest/azure/rp.go:
```diff
@@ -0,0 +1,188 @@
+	}
+
+	// taken from the resources SDK
+	// https://github.com/Azure/azure-sdk-for-go/blob/9f366792afa3e0ddaecdc860e793ba9d75e76c27/arm/resources/resources/providers.go#L252
```
Could you update this to also include an explanation of why we had to Copy/Paste?

mcardosos replied:
Fixed

marstr replied:
Hmmm, I'm not sure I like the updated comment any better. What I'd like to see expressed is that we can't/don't want to take a dependency on the github.com/Azure/azure-sdk-for-go repository. For that reason, we must have a manually maintained copy here, and it made sense to just grab what we had generated.

marstr commented on autorest/azure/rp.go:
```diff
@@ -0,0 +1,188 @@
+		return autorest.SenderFunc(func(r *http.Request) (resp *http.Response, err error) {
+			attempts := 0
+			rr := autorest.NewRetriableRequest(r)
+			for attempts < client.RetryAttempts {
```
[nit] I'd rename `attempts` to something that didn't sound static like `currentAttempt`. Also, it would be a little more readable/conventional to declare `attempts` in the for body like so:

``` Go
for currentAttempt := 0; currentAttempt < client.RetryAttempts; currentAttempt++ {
```

mcardosos replied:
Fixed
